# Assignment  Python  For loop solution 
      Basic Level:

# 1. Write a Python program to print the numbers from 1 to 10 using a `for` loop.

In [1]:
# Using a for loop to iterate over a range of numbers from 1 to 11 (exclusive)
for i in range(1, 11):
    # Printing the current value of the loop variable 'i'
    print(i)


1
2
3
4
5
6
7
8
9
10


# 2. Create a program that calculates the sum of all numbers in a list using a `for` loop.

In [10]:
# Define a list of numbers
numbers = [3, 7, 9, 5, 45]

# Initialize a variable to store the sum
sum_result = 0

# Using a for loop to iterate over each element in the list
for num in numbers:
    # Adding the current number to the sum
    sum_result += num

# Printing the result
print("Sum:", sum_result)


Sum: 69


In this program, a list of numbers is defined, a sum variable is 
initialized to zero, and a for loop is used to iterate over the
list, adding each number to the sum. Finally, the total sum is
printed to the console.

# 3. Write a program to print the characters of a string in reverse order using a `for` loop.

In [17]:
# define the string 
my_string="mohammad,shahanshah!"
#use a for loop to iterate over the string in reverse
for char in reversed(my_string):
    #print each charecter
    print(char,end ="")

!hahsnahahs,dammahom


This Python program prints the characters of the string 
"mohammad,shahanshah!" in reverse order using a for loop and the
reversed() function. It iterates through each character in 
reverse and prints them on the same line, resulting in the 
output: !!hahsnahahs,dammahom

# 4. Develop a program that finds the factorial of a given number using a `for` loop.

In [41]:
# Function to calculate the factorial of a number
def factorial(n):
    result = 1

# Using a for loop to multiply numbers from 1 to n
    for i in range(1, n + 1):
        result *= i

    return result

# Input: The number for which factorial is to be calculated
number = 7

# Output: Print the factorial of the input number
print(f"The factorial of {number} is {factorial(number)}")


The factorial of 7 is 5040



This Python program calculates the factorial of a given number using a for loop. It defines a function factorial that multiplies numbers from 1 to the given number. The example input is 5, and the output is "The factorial of 7 is 5040


# 5. Create a program to print the multiplication table of a given number using a `for` loop.

In [48]:
# Input: The number for which the multiplication table is to be printed
num = 13

# Using a for loop to iterate over the range 1 to 11 (exclusive)
for i in range(1, 6):
    # Multiply the number by the loop variable and print the result
    print(f"{num} x {i} = {num * i}")


13 x 1 = 13
13 x 2 = 26
13 x 3 = 39
13 x 4 = 52
13 x 5 = 65


# 6. Write a program that counts the number of even and odd numbers in a list using a `for` loop.

In [2]:
def count_even_odd(numbers):
    even_count = 0
    odd_count = 0

    for number in numbers:
        if number % 2 == 0:
            even_count += 1
        else:
            odd_count += 1

    return even_count, odd_count

# Example usage:
numbers_list = [1, 2, 3, 4, 5, 6,8, 7, 8, 9, 10,9,76,9,8,87]
even_count, odd_count = count_even_odd(numbers_list)

print("Even numbers:", even_count)
print("Odd numbers:", odd_count)


Even numbers: 8
Odd numbers: 8


count_even_odd, which takes a list of numbers as input and counts the number of even and odd numbers using a for loop. Inside the loop, each number is checked for evenness, and the corresponding count is incremented. The final counts are then returned by the function. The example usage at the end demonstrates how to use the function with a sample list and prints the counts of even and odd numbers.







# 7. Develop a program that prints the squares of numbers from 1 to 5 using a `for` loop.

In [8]:
for number in range(1,6):
    square =number**2
    print (f"The square of number{number} is :{square}")

The square of number1 is :1
The square of number2 is :4
The square of number3 is :9
The square of number4 is :16
The square of number5 is :25


This Python program uses a for loop to iterate over numbers from 1 to 5 (inclusive). Inside the loop, it calculates the square of each number using the ** operator and prints the result. The program showcases a simple use of a for loop to perform a repetitive task for a specified range of values

# 8. Create a program to find the length of a string without using the `len()` function.

In [2]:
def find_length(input_string):
    count = 0
    for char in input_string:
        count += 1
    return count

# Example usage:
user_input = input("Enter a string: ")
length = find_length(user_input)
print("Length of the string:", length)


Enter a string: shahanshah
Length of the string: 10


Certainly! This program calculates the length of a string without using the len() function. It iterates through each character in the string using a for loop and increments a counter for each character. The final counter value represents the length of the string.

# 9. Write a program that calculates the average of a list of numbers using a `for` loop.

In [8]:
def calculate_average(number):
    total=0
    count=0
    
    for num in number :
        total+=num
        count+=1
    if count==0:
            print("cannot calculate average for an enpty  list.")
    else:

        average =total/count
        print("Average:", average)
           # Example usage:
number_list = [55,7, 10, 15, 20, 25 ,12]
calculate_average(number_list) 

Average: 20.571428571428573



This Python program uses a for loop to calculate the average of a list of numbers. It iterates through the list, summing up the numbers and counting them. The average is then calculated by dividing the total sum by the count of numbers. If the list is empty, it prints a message stating that the average cannot be calculated for an empty list.

# 10. Develop a program that prints the first `n` Fibonacci numbers using a `for` loop.

In [9]:
def generate_fibonacci(n):
    fibonacci_numbers = [0, 1]

    for i in range(2, n):
        next_number = fibonacci_numbers[i - 1] + fibonacci_numbers[i - 2]
        fibonacci_numbers.append(next_number)

    return fibonacci_numbers

# Example usage:
n = int(input("Enter the value of n: "))
fibonacci_sequence = generate_fibonacci(n)
print(f"The first {n} Fibonacci numbers are: {fibonacci_sequence}")


Enter the value of n: 5
The first 5 Fibonacci numbers are: [0, 1, 1, 2, 3]


This Python program uses a for loop to generate and print the first n Fibonacci numbers. It starts with the initial sequence of 0 and 1, and in each iteration of the loop, it calculates the next Fibonacci number by summing the last two. The resulting sequence is then printed.







# Type:2 ,Intermediate Level:

# 11. Write a program to check if a given list contains any duplicates using a `for` loop.

In [11]:
def has_duplicates(input_list):
    seen = set()

    for item in input_list:
        if item in seen:
            return True
        seen.add(item)

    return False

# Example usage:
user_list = [1, 2, 3, 4, 5, 1]  # This list contains duplicates
if has_duplicates(user_list):
    print("The list contains duplicates.")
else:
    print("The list does not contain duplicates.")


The list contains duplicates.


his program uses a set (seen) to keep track of unique elements encountered so far. It iterates through the list and checks if each element has been seen before. If it finds a duplicate, it returns True; otherwise, it returns False. In the example, my_list contains a duplicate (the number 1), so the program prints "Has duplicates: True".

# 12. Create a program that prints the prime numbers in a given range using a `for` loop.

In [15]:
def is_prime(num):
    if num < 2:
        return False
    for i in range(2, int(num**0.5) + 1):
        if num % i == 0:
            return False
    return True

def print_primes_in_range(start, end):
    for number in range(start, end + 1):
        if is_prime(number):
            print(number)

# Example usage:
start_range = 0
end_range = 19
print(f"Prime numbers between {start_range} and {end_range}:")
print_primes_in_range(start_range, end_range)


Prime numbers between 0 and 19:
2
3
5
7
11
13
17
19


This Python program uses a for loop to iterate through a given range. It defines a function is_prime to check if a number is prime and another function print_primes_in_range to print prime numbers within the specified range. The loop calls the is_prime function for each number and prints the prime numbers. Adjust the range by modifying the start_range and end_range variables as needed.

# 13. Develop a program that counts the number of vowels in a string using a `for` loop.

In [16]:
def count_vowels(input_string):
    vowels = "aeiouAEIOU"
    vowel_count = 0
    for char in input_string:
        if char in vowels:
            vowel_count += 1
    return vowel_count

# Example usage:
input_str = "Hello, World!"
result = count_vowels(input_str)
print(f"The number of vowels in '{input_str}' is: {result}")


The number of vowels in 'Hello, World!' is: 3


In this program, the count_vowels function takes a string as input and iterates through each character using a for loop. It checks if each character is a vowel (both lowercase and uppercase) and increments the vowel_count accordingly. The example usage shows how to use the function with a specific string. Adjust the input_str variable for different input strings.







# 14. Write a program to find the maximum element in a 2D list using a nested `for` loop.

In [20]:
def find_max_element(matrix):
    max_element = float('-inf')
    for row in matrix:
        for element in row:
            if element > max_element:
                max_element = element
    return max_element

# Example usage:
my_2d_list = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

max_value = find_max_element(my_2d_list)
print("The maximum element in the 2D list is:", max_value)


The maximum element in the 2D list is: 9


this Python program utilizes nested for loops to iterate through each element in a 2D list. The function find_max_element initializes a variable max_element to negative infinity and updates it whenever a larger element is encountered. The example usage with a 2D list demonstrates finding the maximum element. Adjust the input matrix (my_2d_list) for different scenarios.

# 15. Create a program that removes all occurrences of a specific element from a list using a `for` loop.


In [24]:
def remove_element(lst, element):
    for item in lst[:]:
        if item == element:
            lst.remove(item)

# Example usage:
my_list = [1, 2, 3, 4, 2, 5, 2]
remove_element(my_list, 2)
print("List after removing all occurrences of 2:", my_list)


List after removing all occurrences of 2: [1, 3, 4, 5]


This program uses a for loop to iterate through a list and removes all occurrences of a specific element (2 in t

# 16. Develop a program that generates a multiplication table for numbers from 1 to 5 using a nested `for` loop.

In [26]:
for i in range(1, 5):
    for j in range(1, 6):
        print(i * j, end="\t")
    print()


1	2	3	4	5	
2	4	6	8	10	
3	6	9	12	15	
4	8	12	16	20	


This program uses nested for loops to generate a multiplication table for numbers from 1 to 5.



# 17. Write a program that converts a list of Fahrenheit temperatures to Celsius using a `for` loop.

In [34]:
def fahrenheit_to_celsius(fahrenheit_temps):
    celsius_temps = [(temp - 32) * 5/9 for temp in fahrenheit_temps]
    return celsius_temps

# Example usage:
fahrenheit_temps = [32, 68, 104, 212]
celsius_temps = fahrenheit_to_celsius(fahrenheit_temps)
print("Celsius temperatures:", celsius_temps)


Celsius temperatures: [0.0, 20.0, 40.0, 100.0]


This program converts a list of Fahrenheit temperatures to Celsius using a list comprehension within a for loop.

# 18. Create a program to print the common elements from two lists using a `for` loop.

In [35]:
list1 = [1, 2, 3, 4, 5]
list2 = [3, 4, 5, 6, 7]

common_elements = []

for element in list1:
    if element in list2:
        common_elements.append(element)

print("Common elements:", common_elements)


Common elements: [3, 4, 5]


# 19. Develop a program that prints the pattern of right-angled triangles using a `for` loop. Use ‘*’ to draw the pattern

In [36]:
rows = 5

for i in range(1, rows + 1):
    for j in range(i):
        print('*', end=' ')
    print()


* 
* * 
* * * 
* * * * 
* * * * * 


In [37]:
for i in range(1, 6):
    print('* ' * i)


* 
* * 
* * * 
* * * * 
* * * * * 


# 20. Write a program to find the greatest common divisor (GCD) of two numbers using a `for` loop.

In [38]:
def find_gcd(num1, num2):
    gcd = 1
    for i in range(1, min(num1, num2) + 1):
        if num1 % i == 0 and num2 % i == 0:
            gcd = i
    return gcd

# Example usage:
number1 = 24
number2 = 36
result = find_gcd(number1, number2)
print(f"The GCD of {number1} and {number2} is: {result}")


The GCD of 24 and 36 is: 12



This Python program uses a for loop to iteratively find the greatest common divisor (GCD) of two numbers. The find_gcd function checks for common divisors within the range of 1 to the minimum of the two numbers and updates the GCD accordingly. The example usage illustrates finding the GCD of 24 and 36. Adjust input values as needed.

# type. 3:Advanced Level:

# 21. Create a program that calculates the sum of the digits of numbers in a list using a list comprehension.

In [39]:
def digit_sum(number):
    return sum(int(digit) for digit in str(number))

# Example usage:
numbers = [123, 456, 789]
sums = [digit_sum(num) for num in numbers]

print("Original Numbers:", numbers)
print("Sum of Digits:", sums)


Original Numbers: [123, 456, 789]
Sum of Digits: [6, 15, 24]


In this program, the digit_sum function takes a number, converts it to a string, and uses a list comprehension to sum the individual digits. The example usage shows how to apply this function to a list of numbers, calculating the sum of digits for each number in the list. Adjust the numbers list for different input sets.







# 22. Write a program to find the prime factors of a given number using a `for` loop and list comprehension.

In [40]:
def find_prime_factors(number):
    return [factor for factor in range(2, number + 1) if number % factor == 0 and all(factor % i != 0 for i in range(2, int(factor**0.5) + 1))]

# Example usage:
given_number = 56
prime_factors = find_prime_factors(given_number)

print(f"The prime factors of {given_number} are:", prime_factors)


The prime factors of 56 are: [2, 7]


This Python program uses a for loop and list comprehension to find the prime factors of a given number. The find_prime_factors function iterates through the range of potential factors and checks if each factor is prime. The list comprehension collects the prime factors, and the example usage illustrates finding the prime factors of the number 56. Adjust the given_number variable for different input values.







# 23. Develop a program that extracts unique elements from a list and stores them in a new list using a list comprehension.

In [41]:
def extract_unique_elements(input_list):
    return list(set(input_list))

# Example usage:
original_list = [1, 2, 2, 3, 4, 4, 5]
unique_elements = extract_unique_elements(original_list)

print("Original List:", original_list)
print("Unique Elements:", unique_elements)


Original List: [1, 2, 2, 3, 4, 4, 5]
Unique Elements: [1, 2, 3, 4, 5]


This Python program defines a function, extract_unique_elements, which takes a list, converts it to a set to retain only unique elements, and then converts it back to a list. The program demonstrates using this function to extract unique elements from an original list and store them in a new list. Adjust the original_list variable for different input lists.

# 24. Create a program that generates a list of all palindromic numbers up to a specified limit using a list comprehension.


In [47]:
def is_palindrome(number):
    return str(number) == str(number)[::-1]

def generate_palindromic_numbers(limit):
    return [num for num in range(limit + 1) if is_palindrome(num)]

# Example usage:
limit = 100
palindromic_numbers = generate_palindromic_numbers(limit)

print(f"Palindromic numbers up to {limit}: {palindromic_numbers}")


Palindromic numbers up to 100: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 22, 33, 44, 55, 66, 77, 88, 99]


This Python program generates a list of palindromic numbers up to a specified limit using a list comprehension. The is_palindrome function checks if a number is a palindrome, and the generate_palindromic_numbers function utilizes this to create a list of palindromic numbers. The example usage demonstrates generating palindromic numbers up to a limit of 100. Adjust the limit variable for different maximum values.

# 25. Write a program to flatten a nested list using list comprehension.

In [59]:
def flatten_nested_list(nested_list):
    return [element for sublist in nested_list for element in (sublist if isinstance(sublist, list) else [sublist])]

# Example usage:
nested_list = [1, [2, 3], [4, [5, 6]], 7]
flattened_list = flatten_nested_list(nested_list)

print("Original Nested List:", nested_list)
print("Flattened List:", flattened_list)


Original Nested List: [1, [2, 3], [4, [5, 6]], 7]
Flattened List: [1, 2, 3, 4, [5, 6], 7]


This Python program uses a list comprehension to flatten a nested list. The flatten_nested_list function iterates through the elements of the nested list, and if an element is a sublist, it is further iterated to include its elements. The example usage shows flattening a nested list, and you can adjust the nested_list variable for different input lists.

# 26. Develop a program that computes the sum of even and odd numbers in a list separately using list comprehension


In [60]:
def sum_even_odd(numbers):
    even_sum = sum(num for num in numbers if num % 2 == 0)
    odd_sum = sum(num for num in numbers if num % 2 != 0)
    return even_sum, odd_sum

# Example usage:
number_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_sum, odd_sum = sum_even_odd(number_list)

print("Original List:", number_list)
print("Sum of Even Numbers:", even_sum)
print("Sum of Odd Numbers:", odd_sum)


Original List: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Sum of Even Numbers: 30
Sum of Odd Numbers: 25


In this program, the sum_even_odd function takes a list of numbers and uses two list comprehensions to compute the sum of even and odd numbers separately. The sum function is then used to find the total sum for each category. The example usage demonstrates computing the sum of even and odd numbers in a given list. Adjust the number_list variable for different input lists.







# 27. Create a program that generates a list of squares of odd numbers between 1 and 10 using list comprehension


In [61]:
# Using list comprehension to generate squares of odd numbers
squares_of_odd_numbers = [x**2 for x in range(1, 11) if x % 2 != 0]

# Print the result
print("List of Squares of Odd Numbers:", squares_of_odd_numbers)


List of Squares of Odd Numbers: [1, 9, 25, 49, 81]


In this program, the list comprehension iterates through numbers from 1 to 10, and for each odd number, it calculates and includes its square in the resulting list. The if x % 2 != 0 condition ensures that only odd numbers are considered.

# 28. Write a program that combines two lists into a dictionary using list comprehension.

In [62]:
# Example lists
keys = ['name', 'age', 'city']
values = ['Alice', 25, 'New York']

# Combining lists into a dictionary using list comprehension
combined_dict = {keys[i]: values[i] for i in range(min(len(keys), len(values)))}

# Print the result
print("Combined Dictionary:", combined_dict)


Combined Dictionary: {'name': 'Alice', 'age': 25, 'city': 'New York'}


n this program, the list comprehension combines corresponding elements from two lists (keys and values) into key-value pairs in a dictionary. The min(len(keys), len(values)) ensures that the loop iterates up to the length of the shorter list to prevent index errors. Adjust the keys and values lists for different input data.







# 29. Develop a program that extracts the vowels from a string and stores them in a list using list comprehension.

In [63]:
# Function to extract vowels from a string
def extract_vowels(input_string):
    vowels = [char for char in input_string if char.lower() in 'aeiou']
    return vowels

# Example usage:
input_str = "Hello, World!"
vowels_list = extract_vowels(input_str)

print("Original String:", input_str)
print("List of Vowels:", vowels_list)


Original String: Hello, World!
List of Vowels: ['e', 'o', 'o']


In this program, the extract_vowels function uses a list comprehension to iterate through the characters of the input string, checking if each character (converted to lowercase) is a vowel. The example usage demonstrates extracting vowels from the string "Hello, World!". Adjust the input_str variable for different input strings

# 30. Create a program that removes all non-numeric characters from a list of strings using list comprehension.

In [64]:
# Function to remove non-numeric characters from a list of strings
def remove_non_numeric(strings_list):
    numeric_strings = [''.join(char for char in string if char.isdigit()) for string in strings_list]
    return numeric_strings

# Example usage:
string_list = ["abc123", "45xyz", "789!@#"]

numeric_strings_list = remove_non_numeric(string_list)

print("Original List of Strings:", string_list)
print("List of Numeric Strings:", numeric_strings_list)


Original List of Strings: ['abc123', '45xyz', '789!@#']
List of Numeric Strings: ['123', '45', '789']


In this program, the remove_non_numeric function uses a list comprehension to iterate through each string in the input list. It then joins only the numeric characters using char.isdigit(). The example usage demonstrates removing non-numeric characters from a list of strings. Adjust the string_list variable for different input lists.

# Type 4:Challenge Level:

# 31. Write a program to generate a list of prime numbers using the Sieve of Eratosthenes algorithm and list comprehension.

In [1]:
def sieve_of_eratosthenes(limit):
    # Create a boolean list "is_prime[0..limit]" and initialize all entries it as true.
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False  # 0 and 1 are not prime numbers

    # Iterate through all numbers starting from 2 (the first prime)
    for num in range(2, int(limit**0.5) + 1):
        if is_prime[num]:
            # Mark all multiples of num as not prime
            for multiple in range(num**2, limit + 1, num):
                is_prime[multiple] = False

    # Generate the list of prime numbers
    primes = [num for num in range(2, limit + 1) if is_prime[num]]
    return primes

# Example: Generate prime numbers up to 50
limit = 50
prime_numbers = sieve_of_eratosthenes(limit)
print("Prime numbers up to", limit, "are:", prime_numbers)


Prime numbers up to 50 are: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]


Function Definition: The program defines a function sieve_of_eratosthenes that takes a limit as input.

Initialize Boolean List: It creates a boolean list is_prime of size (limit + 1) and initializes all entries as True. It then marks 0 and 1 as False since they are not prime.

Sieve Process: It iterates through numbers starting from 2 up to the square root of the limit. For each prime number found, it marks its multiples as not prime.

Generate Prime List: After the sieve process, the program creates a list of prime numbers from the boolean list where the value is still True.

Example Usage: The program demonstrates its usage by generating prime numbers up to a specified limit (in this case, 50) and prints the result.

# 32. Create a program that generates a list of all Pythagorean triplets up to a specified limit using listcomprehension.
.

In [ ]:
def generate_pythagorean_triplets(limit):
    triplets = [(a, b, c) for a in range(1, limit + 1) for b in range(a, limit + 1) for c in range(b, limit + 1) if a**2 + b**2 == c**2]
    return triplets

# Specify the limit
limit = int(input("Enter the limit for Pythagorean triplets: "))

# Generate and print the Pythagorean triplets
triplets = generate_pythagorean_triplets(limit)
print("Pythagorean Triplets up to {}: {}".format(limit, triplets))


Enter the limit for Pythagorean triplets: 70704015044


This program uses list comprehension to generate Pythagorean triplets (a, b, c) where a, b, and c are integers up to the specified limit. The condition a**2 + b**2 == c**2 ensures that only valid Pythagorean triplets are included in the list. The triplets are then printed to the console.







# 33. Develop a program that generates a list of all possible combinations of two lists using list comprehension.

In [2]:
list1 = [1, 2, 3]
list2 = ['a', 'b', 'c']

combinations = [(x, y) for x in list1 for y in list2]
print("All possible combinations of {} and {}: {}".format(list1, list2, combinations))



All possible combinations of [1, 2, 3] and ['a', 'b', 'c']: [(1, 'a'), (1, 'b'), (1, 'c'), (2, 'a'), (2, 'b'), (2, 'c'), (3, 'a'), (3, 'b'), (3, 'c')]



The provided Python program uses list comprehension to generate all possible combinations of elements from two given lists. It defines two lists (list1 and list2), and then uses a list comprehension to create a new list (combinations) containing tuples of elements from the two lists. Finally, it prints the original lists and the generated combinations.

# 34. Write a program that calculates the mean, median, and mode of a list of numbers using list comprehension.

In [11]:
from statistics import mean, median, mode

# Input list of numbers
numbers = [float(x) for x in input("Enter a list of numbers separated by spaces: ").split()]

# Calculate mean, median, and mode
mean_value = mean(numbers)
median_value = median(numbers)
try:
    mode_value = mode(numbers)
except StatisticsError:
    mode_value = "No unique mode"

# Display results
print("Mean: {:.2f}".format(mean_value))
print("Median: {:.2f}".format(median_value))
print("Mode: {}".format(mode_value))


Enter a list of numbers separated by spaces: 6 7 4 6
Mean: 5.75
Median: 6.00
Mode: 6.0


This program uses the statistics module to calculate the mean, median, and mode of a list of numbers. It takes user input for the list of numbers, converts them to float using list comprehension, and then calculates the mean, median, and mode. The results are then displayed.

You can run this program, enter a list of numbers when prompted, and it will calculate and display the mean, median, and mode for the given list.

# 35. Create a program that generates Pascal's triangle up to a specified number of rows using list comprehension.


In [ ]:
def generate_pascals_triangle(rows):
    triangle = [[1] if i == 0 else [1] + [triangle[i-1][j-1] + triangle[i-1][j] for j in range(1, i)] + [1] for i in range(rows)]
    return triangle

# Input for the number of rows
rows = int(input("Enter the number of rows for Pascal's triangle: "))

# Generate and print Pascal's triangle
pascals_triangle = generate_pascals_triangle(rows)
for row in pascals_triangle:
    print(row)


Certainly! The provided Python program uses list comprehension to generate Pascal's triangle up to a specified number of rows. The program defines a function, generate_pascals_triangle, which uses list comprehension to construct each row of Pascal's triangle based on the elements from the previous row. The user is prompted to input the number of rows, and the program then prints Pascal's triangle up to the specified number of rows

# 36. Develop a program that calculates the sum of the digits of a factorial of numbers from 1 to 5 using list
comprehension.

In [1]:
import math

# Calculate the factorial of a number
def factorial(n):
    return math.factorial(n)

# Calculate the sum of the digits of a number
def sum_of_digits(number):
    return sum(int(digit) for digit in str(number))

# Calculate the sum of the digits of the factorial of numbers from 1 to 5
sums = [sum_of_digits(factorial(n)) for n in range(1, 6)]

# Print the results
for n, s in zip(range(1, 6), sums):
    print(f"The sum of the digits of {n}! is: {s}")


The sum of the digits of 1! is: 1
The sum of the digits of 2! is: 2
The sum of the digits of 3! is: 6
The sum of the digits of 4! is: 6
The sum of the digits of 5! is: 3


In this program, the factorial function from the math module calculates the factorial of a given number. The sum_of_digits function calculates the sum of the digits of a given number by converting it to a string and summing its integer digits. The list comprehension generates a list of sums for the factorial of numbers from 1 to 5. Finally, the program prints the results.







# 37. Write a program that finds the longest word in a sentence using list comprehension.

In [2]:
# Input sentence
sentence = input("Enter a sentence: ")

# Find the longest word using list comprehension
longest_word = max((word.strip(".,!?") for word in sentence.split()), key=len)

# Display the result
print(f"The longest word in the sentence is: {longest_word}")


Enter a sentence: mohammad shahahanshah
The longest word in the sentence is: shahahanshah


The Python program uses list comprehension to find the longest word in a user-input sentence. It splits the sentence into words, removes punctuation, and then identifies the longest word by applying the max function with the key set to the length of each word. The result is then displayed to the user.

# 38. Create a program that filters a list of strings to include only those with more than three vowels using list comprehension.


In [4]:
# Function to count vowels in a string
def count_vowels(word):
    return sum(1 for char in word.lower() if char in 'aeiou')

# Input list of strings
word_list = input("Enter a list of strings separated by spaces: ").split()

# Filter words with more than three vowels using list comprehension
filtered_words = [word for word in word_list if count_vowels(word) > 3]

# Display the result
print("Words with more than three vowels: {}".format(filtered_words))


Enter a list of strings separated by spaces: shhaha moham kasjd
Words with more than three vowels: []


Count Vowels Function: The count_vowels function is defined to count the number of vowels in a given word. It converts the word to lowercase and uses list comprehension to count the vowels.

User Input: The user is prompted to enter a list of strings separated by spaces.

List Comprehension: The program uses list comprehension to create a new list (filtered_words). It iterates over each word in the input list and includes only those words for which the count of vowels (using count_vowels) is greater than 3.

Display Result: Finally, the program prints the words from the list that have more than three vowels.

# 39. Develop a program that calculates the sum of the digits of numbers from 1 to 1000 using list comprehension.


In [5]:
# Function to calculate the sum of digits in a number
def sum_of_digits(number):
    return sum(int(digit) for digit in str(number))

# Calculate the sum of digits for numbers from 1 to 1000 using list comprehension
sums = [sum_of_digits(n) for n in range(1, 1001)]

# Print the results
print(f"The sum of digits for numbers from 1 to 1000 is: {sum(sums)}")


The sum of digits for numbers from 1 to 1000 is: 13501


The sum_of_digits function calculates the sum of digits in a given number by converting it to a string and summing its integer digits.

List comprehension is used to generate a list (sums) of the sum of digits for numbers from 1 to 1000.

The program prints the total sum of digits for all numbers in the specified range.

You can run this program to find the sum of digits for numbers from 1 to 1000.

# 40. Write a program that generates a list of prime palindromic numbers using list comprehension.

In [7]:
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Generate a list of prime palindromic numbers using list comprehension
prime_palindromes = [num for num in range(1, 1000) if is_prime(num) and str(num) == str(num)[::-1]]

# Display the result
print("Prime palindromic numbers: {}".format(prime_palindromes))


Prime palindromic numbers: [2, 3, 5, 7, 11, 101, 131, 151, 181, 191, 313, 353, 373, 383, 727, 757, 787, 797, 919, 929]


This Python program generates a list of prime palindromic numbers using list comprehension. It defines a function (is_prime) to check if a number is prime, and then uses list comprehension to create a list of numbers from 1 to 999 that are both prime and palindromic. The resulting list is then displayed.

# THANK YOU